### PR TITLE
Fix #1673 - safety inference in tree structure of REST API

### LINKED ISSUE #1673: vibe-0.8.0-beta.2 - tree structure in REST API
If I have a tree structure defined in REST API newly annotated with @safe it fails to register such an API.

Try this:

```D
@safe unittest {
	static struct Foo { Foo[] foos; }
	serialize!TestSerializer(Foo());
}
```
It ends up with
```
data/vibe/data/serialization.d(1588,26): Error: @safe function 'vibe.data.serialization.__unittestL1586_178' cannot call @system function 'vibe.data.serialization.serialize!(TestSerializer, Foo).serialize'
```

The problem seems to be with safety inference here: https://github.com/rejectedsoftware/vibe.d/blob/master/data/vibe/data/serialization.d#L340

Not sure what to do about it.

### PR DESCRIPTION
**Pull request by tchaloupka**
I solved it by annotating serializeValue method and fixed all unittests so it compiles (with dmd-2.073.0).

There are some more pragma deprecation messages that can be fixed, but I wanted this to be minimal and to see if this is even acceptable.

### CODE REVIEW

**Comment by s-ludwig** [comment]
I'm kind of torn on this one and #1671, because on one hand the reasons for both are very valid (in case of the error messages IMO DMD really should generally output the most specific place where `@safe`/`@nogc`/`nothrow` inference failed). On the other hand, forcefully adding `@safe`/`@trusted` breaks inference completely, so now either all user defined serialization methods need to be annotated safe, or the library opens a hidden `@trusted` safety hole.

I will have a look if I can find a higher level spot where the REST code can be fixed.

**Comment by tchaloupka** [comment]
Problem here is worse in that even the struct is safe the inference doesn't work because of cycle of the tree structure which is infered as system as specified in https://dlang.org/spec/function.html#function-attribute-inference

**Comment by s-ludwig** [comment]
Interesting, good to know. Do you possibly know if the decision to infer recursive functions as `@system` was discussed somewhere? It seems a bit arbitrary at first sight.

**Comment by tchaloupka** [comment]
I don't remember nor found any discussion about this. Just somewhat related bug report: https://issues.dlang.org/show_bug.cgi?id=16528

**Comment by tchaloupka** [comment]
Any news about this?
Currently (0.8.0-beta.3) it is impossible to make REST API with tree structure

```D
struct Foo
{
    Foo[] foos;
}
```

**Comment by tchaloupka** [comment]
I've rebased this so now it's evens simpler.
This is stays to be a problem when tree structures are used in REST API.

**Comment by tchaloupka** [comment]
I've resolved the conflict, retested it and somehow @s-ludwig probably fixed this by other change because now it compiles fine without any safety attribute change. (tried with dmd-2.074.0 and dmd-2.073.2).

So this was reduced only to the unittest for the tree structure.. :)

**Comment by tchaloupka** [comment]
Rebased again. Thought that it now works without safe addition, but forgot to unittest with `--combined` switch..

File: data/vibe/data/serialization.d [hunk 1]
```diff
@@ -1474,14 +1474,14 @@ unittest // Testing corner case: Variadic template constructors and methods
 @safe unittest // Immutable data deserialization
 {
 	import vibe.data.json;
-
```
Comment (wilzbach) [inline data/vibe/data/serialization.d]: Unrelated changes due to trailing whitespace - this is the past!
-> You will like https://github.com/rejectedsoftware/vibe.d/pull/1800 :)

**Comment by s-ludwig** [comment]
There is one case that I would like to avoid, namely that passing a `@system`  output range for serialization fails to compile. I'll pull this to master now and will then see if it's possible to use a string mixin to fake the job of the compiler here. If that works out, I'll squeeze this into 0.8.0, because it is indeed a regression.

**Pull request merged**